tasks/framework-tools/src/utils.js

```diff
@@ -0,0 +1,238 @@
+const terminalLink = require('terminal-link')
+
+const REDWOOD_PACKAGES_PATH = path.resolve(__dirname, '../../../packages')
+
+function sortObjectKeys(obj) {
+  return Object.keys(obj)
+    .sort()
+    .reduce((acc, key) => {
+      acc[key] = obj[key]
+      return acc
+    }, {})
+}
+
+/**
+ * The Redwood packages that are installed into a Redwood Project.
+ */
+function frameworkPackages() {
+  return fg.sync('**/package.json', {
+    cwd: REDWOOD_PACKAGES_PATH,
+    ignore: ['**/node_modules/**', '**/create-redwood-app/**'],
+    absolute: true,
+  })
+}
+
+/**
+ * The dependencenies used by RedwoodJS packages.
+ */
+function gatherDeps(packages = frameworkPackages()) {
+  const warnings = []
+  const dependencies = {}
+
+  for (const packageFile of packages) {
+    // try catch to handle parsing errors in package json
+    try {
+      const packageJson = JSON.parse(fs.readFileSync(packageFile))
+      for (const [name, version] of Object.entries(
+        packageJson?.dependencies ?? {}
+      )) {
+        // Skip `@redwoodjs/*` packages, since these are processed
+        // by the workspace.
+        if (!name.startsWith('@redwoodjs/')) {
+          if (dependencies[name] && dependencies[name] !== version) {
+            warnings.push([
+              name,
+              'dependency version mismatched, please make sure the versions are the same.',
+            ])
+          }
+          dependencies[name] = version
+        }
+      }
+    } catch (error) {
+      console.error()
+      console.error(c.red(`Error in ${packageFile}:`))
+      console.error('', error.message)
+      console.error()
+      console.error(` This package's dependencies will not be included.`)
+      console.error()
+    }
+  }
+  return { dependencies: sortObjectKeys(dependencies), warnings }
+}
+
+function logWarnings(warnings) {
+  for (const [packageName, message] of warnings) {
+    console.warn(c.yellow('Warning:'), packageName, message)
+  }
+}
+
+/**
+ * The files that are published by the Redwood Framework.
+ * Note: The packages must be built.
+ */
+function packagesFileList(packages = frameworkPackages()) {
+  const fileList = {}
+  for (const packageFile of packages) {
+    const packageJson = require(packageFile)
+
+    if (!packageJson.name) {
+      continue
+    }
+
+    fileList[packageJson.name] = packlist.sync({
+      path: path.dirname(packageFile),
+    })
+  }
+  return fileList
+}
+
+function redwoodBins(packages = frameworkPackages()) {
+  let bins = {}
+  for (const packageFile of packages) {
+    const packageJson = require(packageFile)
+    if (!packageJson.name) {
+      continue
+    }
+
+    if (!packageJson.bin) {
+      continue
+    }
+    for (const [binName, binPath] of Object.entries(packageJson.bin)) {
+      bins[binName] = path.join(packageJson.name, binPath)
+    }
+  }
+  return bins
+}
+
+// gets the redwood project's package.json and builds a write function around it
+function getPackageJson(projectPath) {
```

Comment:
I actually think that this abstraction and encapsulation of the package.json file makes it more difficult to reason about, and would be a good source of technical debt if we wanted to modify more than just the dependencies. As an example we may want to update the devDependencies, and then we'll have to add a new function.

On the surface we're doing 3 things:

1. Reading the `package.json` file: 
```js
const p = JSON.parse(fs.readFileSync('/path/to/package.json'))
```
2. Modifying it:
```js
p.depdencies = {
 ...p.dependencies,
 newDeps,
}
```
3. And writing those to the same location:
```js
fs.writeFileSync(JSON.stringify(p, undefined, 2))
```

Which is just 5 lines of code, maybe this isn't worth it?

Reply:
Maybe this can just be a single function: `frameworkDependenciesToProject`

Reply:
> would be a good source of technical debt if we wanted to modify more than just the dependencies

We can pass any draft function to immer's produce function. The one for updating deps can be the default, or we can make functions that we can pass to make things more composable

Reply:
Not trying to be argumentative, but I feel like the only extra things this function does than the three things you enumerated are 1) make a terminal link to the file and 2) handle errors. Is it just not worth it make those two things repeatable? I had to handle both for copying deps and syncing

```js
  const packageJsonLink = terminalLink(
    'Redwood Project Path',
    'file://' + packageJsonPath
  )

  if (!fs.existsSync(packageJsonPath)) {
    console.log(
      `Error: The specified ${packageJsonLink} does not have a package.json file.`
    )
    console.log('Expected:', packageJsonPath)
    process.exit(1)
  }

```